### ISSUE TITLE
Possible bug on checking the size parameter in the function sd_ndisc_router_from_raw

### ISSUE THREAD

**Issue by liftdat** [systemd/systemd#23258]
In the file `ndisc-router.c`, the function `sd_ndisc_router_from_raw` has the following [code](https://github.com/systemd/systemd/blob/fa2ba7aea87ad1812f730f0c2ad86fbfe2ca6552/src/libsystemd-network/ndisc-router.c#L34):

```
int sd_ndisc_router_from_raw(sd_ndisc_router **ret, const void *raw, size_t raw_size) {
    _cleanup_(sd_ndisc_router_unrefp) sd_ndisc_router *rt = NULL;
    int r;

    assert_return(ret, -EINVAL);
    assert_return(raw || raw_size <= 0, -EINVAL);

    rt = ndisc_router_new(raw_size);
}
```

Why the assertion checks whether `raw_size` is less than or equal to zero?  Since  `raw_size` is an unsigned type, this check is actually equivalent to ` raw_size == 0`.

Moreover, if some caller calls the function as: `sd_ndisc_router_from_raw(ret, raw, -1)` , then the checks can be by passed. Inside `ndisc_router_new`,  an integer overflow will happen when computing the allocation size of the buffer, which will lead to subsequent buffer overflow:

```
sd_ndisc_router *ndisc_router_new(size_t raw_size) {
    sd_ndisc_router *rt;

    rt = malloc0(ALIGN(sizeof(sd_ndisc_router)) + raw_size);
    ...
}
```

The function `sd_ndisc_router_from_raw` seems to be public and the assertion in the code suggests the above scenario.  I think either there is a bug in the code or the assertion should be changed to be less misleading.

**Comment by yuwata** [comment]
`raw_size <= 0` is equipment to `raw_size == 0`.

**Comment by evverx** [comment]
Judging by
```sh
$ git grep sd_ndisc_router_from_raw
src/libsystemd-network/ndisc-router.c:int sd_ndisc_router_from_raw(sd_ndisc_router **ret, const void *raw, size_t raw_size) {
src/systemd/sd-ndisc.h:int sd_ndisc_router_from_raw(sd_ndisc_router **ret, const void *raw, size_t raw_size);
```
that function isn't used anywhere though and it hasn't been public since 17347053cf118ea27f5d was merged.

@liftdat I wonder if it was generated by the same static analyzer as https://github.com/systemd/systemd/issues/22873?

**Comment by liftdat** [comment]
> FWIW the code actually used to parse data has been fuzzed on OSS-Fuzz since 2018

If the piece of code is not meant to be used in public, then it is fine. I just think the assertion in the code is misleading.

**Comment by evverx** [comment]
@liftdat I agree that the code is misleading and I'm not sure why it's in the repository.

I'm just trying to understand where bug reports like that come from and what static analyzer is used to generate them

**Comment by yiyuaner** [comment]
> @liftdat I agree that the code is misleading and I'm not sure why it's in the repository.
> 
> I'm just trying to understand where bug reports like that come from and what static analyzer is used to generate them

Hi, it comes from a static analyzer (under research development). It focus on suspicious buffer overflow bugs.

**Comment by evverx** [comment]
As far as I understand it has analyzed the "network" part of systemd so far (because I presume it's the most interesting and impactful target) but I wonder if it was used to scan the other parts of systemd and whether it has found anything there?

**Comment by yiyuaner** [comment]
> As far as I understand it has analyzed the "network" part of systemd so far (because I presume it's the most interesting and impactful target) but I wonder if it was used to scan the other parts of systemd and whether it has found anything there?

We have used it to scan the whole project and reported all the plausible reports after manual examination.

**Comment by evverx** [comment]
Got it. Thanks. Assuming it's academic research It would be interesting to take a look at the paper when/if it comes out.

**Comment by yuwata** [comment]
Fix is waiting in #23260.
